nifi-nar-bundles/nifi-azure-bundle/nifi-azure-graph-authorizer/src/main/java/org/apache/nifi/authorization/azure/AzureGraphUserGroupProvider.java

```diff
@@ -0,0 +1,520 @@
+                        }
+                    } else {
+                        userName = jsonUser.get("userPrincipalName").getAsString();
+                    }
+                    final User user = new User.Builder().identifier(idUser).identity(userName).build();
+                    idUserMap.put(idUser, user);
+                    nameUserMap.put(userName, user);
+                    groupBuilder.addUser(idUser);
+                }
+                IDirectoryObjectCollectionWithReferencesRequestBuilder nextPageRequest = userpage.getNextPage();
+
+                if (nextPageRequest != null) {
+                    userpage = nextPageRequest.buildRequest().get();
+                } else {
+                    break;
+                }
+            }
+            final Group group = groupBuilder.build();
+            return new UserGroupQueryResult(group, idUserMap, nameUserMap);
+        } else {
+            return null;
+
+        }
+    }
+
+    /**
+     * refresh the user & group data for UserGroupProvider plugin service
+     * @param groupDisplayNames a list of group display names
+     */
+    private void refreshUserGroupData(List<String> groupDisplayNames) throws IOException, ClientException {
+        final long startTime = System.currentTimeMillis();
+        // load data into temporary data structure before swapping out
+        final Map<String, User> _usersById = new HashMap<String, User>(); // id == identifier
+        final Map<String, User> _usersByName = new HashMap<String, User>(); // name == identity
+        final Map<String, Group> _groupsById = new HashMap<String, Group>();
+
+        for (String grpFilter : groupDisplayNames) {
+
+            UserGroupQueryResult queryResult = getUsersFrom(grpFilter);
+            if (queryResult != null) {
+                _usersById.putAll(queryResult.idToUserMap);
+                _usersByName.putAll(queryResult.nameToUserMap);
+                _groupsById.put(queryResult.group.getIdentifier(), queryResult.group);
+            }
+        }
+
+        synchronized (usersById) {
+            usersById.clear();
+            usersById.putAll(_usersById);
+        }
+
+        synchronized (usersByName) {
+            usersByName.clear();
+            usersByName.putAll(_usersByName);
+        }
+
+        synchronized (groupsById) {
+            groupsById.clear();
+            groupsById.putAll(_groupsById);
+        }
+        final long endTime = System.currentTimeMillis();
+        logger.info("Refreshed users and groups, took {} miliseconds", (endTime - startTime));
+
+    }
+
+    @Override
+    public void preDestruction() throws AuthorizerDestructionException {
+        try {
+            scheduler.shutdownNow();
+        } catch (final Exception e) {
+            logger.warn("Error shutting down refresh scheduler: " + e.getMessage(), e);
+        }
+    }
+
+    class UserGroupQueryResult {
```

Comment:
This should should users getters instead of public fields; however it might not be needed given comments about immutable user group class.
```suggestion
    private static class UserGroupQueryResult {
```